### PR TITLE
MultiQC support (RNA-seq)

### PR DESCRIPTION
**Pull request by jkokosar**
## Checklist

* [ ] Update CHANGELOG.rst for each commit separately:
  * Pay attention to write entries under the "Unreleased" section.
  * Mark all breaking changes as "**BACKWARD INCOMPATIBLE:**" and put them
    before non-breaking changes.
  * If a commit modifies a feature listed under "Unreleased" section,
    it might be sufficient to modify the existing CHANGELOG entry from previous
    commit(s).
* [ ] Bump the process version:
  * **MAJOR version (first number)**: Backward incompatible changes (changes
    that break the api/interface). Examples: renaming the input/output, adding
    mandatory input, removing input/output...
  * **MINOR version (middle number)**: add functionality or changes in a
    backwards-compatible manner. Examples: add output field, add non-mandatory
    input parameter, use a different tool that produces same results...
  * **PATCH version (last number)**: changes/bug fixes that do not affect
    the api/interface. Examples: typo fix, change/add warning messages...
* [ ] All inputs are used in process.
* [ ] All output fields have their ``re-save`` calls.

### Processes ###

* Code that was once accepted should be kept like it is. It should be changed
  only when its functionality has changed. Small parts may be changed
  to fix bugs (by changing code only to make it nicer you might introduce a bug).
* Set the optimal number of cores that the tool can use (check if parallelization
  can be performed).
* Description of tool should be self explanatory with links to tool's homepage
  and publication.
* Write explanatory input labels and add short command line parameter names
  in square brackets (for example: Maximum length [``--maxlength``]).
* Where possible provide default values for the process input fields.
  If they are provided, then ``required: false`` is not needed.
* Add descriptive error and warning messages.
* Place a ``re-checkrc`` call after each process step (call of a tool, a
  Bash command, a Python script or an R script).
* Add ``re-progress`` calls where appropriate.

### Commit and PR messages ###

* Each commit should be minimal (i.e. 1 change) and self-contained (including
  tests).
* Changes to Docker images need to be accepted in a separate pull request.
  Before the new images can be used for automated testing of your PR, they
  must be manually tagged and pushed to Docker Hub.
* Mark incomplete PRs with the [WIP] (Work In Progress) tag.
* Do not paste links to private repositories from the public one.

### Tests ###

* Add meaningful names of test files with spaces:
  * Genomes: ``genome species.fa.gz``
  * Annotations: ``annotation species.gtf.gz``
  * Adapters: ``adapters-name/type-source``.
* Output and specific input files should include the name of the tested tool.
* Keep test files' size small. If there is no other option, add the large
  test file to the ``files/large`` directory so it will be tracked with Git LFS.
* Pay attention to edge cases: what will happen if any data is
  missing, if length of some list is 0, if string is given where one expects
  integer, if file is empty, ...
* When testing workflows, try to verify workflow success by checking one
  of the last steps. Pick the one that is most likely to fail. Also test
  that all steps have 'OK' status.

### CODE REVIEW

**Comment by mstajdohar** [comment]
Rebase

File: resolwe_bio/tests/workflows/test_rna_seq.py [hunk 1]
```diff
@@ -132,15 +156,49 @@ def test_bbduk_star_featurecounts_workflow(self):
...
 
         inputs['preprocessing']['reads'] = paired_reads.id
         self.run_process('workflow-bbduk-star-featurecounts-paired', inputs)
         for data in Data.objects.all():
             self.assertStatus(data, Data.STATUS_DONE)
-        workflow = Data.objects.last()
-        self.assertFile(workflow, 'rc', 'feature_counts_rc_paired.tab.gz', compression='gzip')
+        feature_counts = Data.objects.filter(process__slug='feature_counts').order_by('-created')[0]
+        self.assertFile(feature_counts, 'rc', 'feature_counts_rc_paired.tab.gz', compression='gzip')
+        multiqc = Data.objects.filter(process__slug='multiqc').order_by('created')[1]
```
Comment (JureZmrzlikar) [inline resolwe_bio/tests/workflows/test_rna_seq.py]: You can use `order_by('-created')[0]`, to be consistent with other cases.

File: resolwe_bio/tests/workflows/test_rna_seq.py [hunk 2]
```diff
@@ -102,20 +102,44 @@ def test_bbduk_star_htseq_paired_workflow(self):
...
                 'genome2': star_index_fasta.id,
             }
             star_index = self.run_process('alignment-star-index', inputs)
             adapters = self.prepare_adapters()
 
+            rrna_reference = self.run_process('upload-fasta-nucl', {
+                'src': 'Homo_sapiens_rRNA.fasta.gz',
+                'source': 'NCBI',
+                'species': 'Homo sapiens',
+                'build': 'rRNA'
```
Comment (JureZmrzlikar) [inline resolwe_bio/tests/workflows/test_rna_seq.py]: Add comma at the end

File: resolwe_bio/tests/workflows/test_rna_seq.py [hunk 3]
```diff
@@ -102,20 +102,44 @@ def test_bbduk_star_htseq_paired_workflow(self):
...
+                fn='hs annotation.gtf.gz',
+                source='ENSEMBL',
+                species='Homo sapiens',
+                build='ens90',
+            )
+            star_index_fasta = self.run_process('upload-fasta-nucl', {
+                'src': 'hs genome.fasta.gz',
+                'source': 'ENSEMBL',
+                'species': 'Homo sapiens',
+                'build': 'ens90'
```
Comment (JureZmrzlikar) [inline resolwe_bio/tests/workflows/test_rna_seq.py]: Add comma

File: resolwe_bio/processes/workflows/bbduk_star_featurecounts_qc.yml [hunk 4]
```diff
@@ -0,0 +1,998 @@
...
+          label: Show advanced parameters
+          type: basic:boolean
+          default: false
+        - name: unstranded
+          label: The data is unstranded
+          type: basic:boolean
+          default: false
+          hidden: '!alignment.show_advanced'
+          description: >
+            For unstranded RNA-seq data, Cufflinks/Cuffdiff require spliced alignments with XS strand attribute, which STAR will generate with --outSAMstrandField intronMotif option. As required, the XS strand attribute will be generated for all alignments that contain splice junctions. The spliced alignments that have undefined strand (i.e. containing only non-canonical unannotated junctions) will be suppressed. If you have stranded RNA-seq data, you do not need to use any specific STAR options. Instead, you need to run Cufflinks with the library option --library-type options. For example, cufflinks --library-type fr-firststrand should be used for the standard dUTP protocol, including Illumina's stranded Tru-Seq. This option has to be used only for Cufflinks runs and not for STAR runs.
```
Comment (JureZmrzlikar) [inline resolwe_bio/processes/workflows/bbduk_star_featurecounts_qc.yml]: This is a very long line. Consider wrapping it.

File: resolwe_bio/processes/workflows/bbduk_star_featurecounts_qc.yml [hunk 5]
```diff
@@ -0,0 +1,998 @@
...
+        - name: detect_chimeric
+          label: Detect chimeric and circular alignments
+          hidden: '!alignment.show_advanced'
+          group:
+            - name: chimeric
+              label: Detect chimeric and circular alignments
+              type: basic:boolean
+              default: false
+              description: >
+                To switch on detection of chimeric (fusion) alignments (in addition to normal mapping), --chimSegmentMin should be set to a positive value. Each chimeric alignment consists of two "segments". Each segment is non-chimeric on its own, but the segments are chimeric to each other (i.e. the segments belong to different chromosomes, or different strands, or are far from each other). Both segments may contain splice junctions, and one of the segments may contain portions of both mates. --chimSegmentMin parameter controls the minimum mapped length of the two segments that is allowed. For example, if you have 2x75 reads and used --chimSegmentMin 20, a chimeric alignment with 130b on one chromosome and 20b on the other will be output, while 135 + 15 won't be.
```
Comment (JureZmrzlikar) [inline resolwe_bio/processes/workflows/bbduk_star_featurecounts_qc.yml]: Wrap line.

File: resolwe_bio/processes/workflows/bbduk_star_featurecounts_qc.yml [hunk 6]
```diff
@@ -0,0 +1,998 @@
...
+        - name: t_coordinates
+          label: Output in transcript coordinates
+          hidden: '!alignment.show_advanced'
+          group:
+            - name: quantmode
+              label: Output in transcript coordinates
+              type: basic:boolean
+              default: false
+              description: >
+                With --quantMode TranscriptomeSAM option STAR will output alignments translated into transcript coordinates in the Aligned.toTranscriptome.out.bam file (in addition to alignments in genomic coordinates in Aligned.*.sam/bam files). These transcriptomic alignments can be used with various transcript quantification software that require reads to be mapped to transcriptome, such as RSEM or eXpress.
```
Comment (JureZmrzlikar) [inline resolwe_bio/processes/workflows/bbduk_star_featurecounts_qc.yml]: Also here and some other cases down the line...

File: resolwe_bio/tests/processes/test_support_processors.py [hunk 7]
```diff
@@ -186,3 +186,54 @@ def test_library_strandedness(self):
...
+                paired_reads.id,
+                filtered_reads.id,
+                bam_samtools.id,
+                star_alignment.id,
+            ],
+            'advanced': {
+                'dirs': True,
+                'config': True,
+            }
+        })
```
Comment (JureZmrzlikar) [inline resolwe_bio/tests/processes/test_support_processors.py]: IS it enough to know that process finished without errors? Or do we want a more strict way of checking the consistency of results?

File: resolwe_bio/tests/workflows/test_rna_seq.py (hunk 1 above)
Comment (JureZmrzlikar) [reply resolwe_bio/tests/workflows/test_rna_seq.py]: Even better, rather use `latest('created')` or `last()`. Know that `last()` will order by primary key (`id`).

File: docs/CHANGELOG.rst [hunk 8]
```diff
@@ -12,21 +12,23 @@ Unreleased
 
 Changed
 -------
-* **BACKWARD INCOMPATIBLE:** Require Resolwe 12.x
-* **BACKWARD INCOMPATIBLE:** Remove obsolete processes: ``assembler-abyss``,
+- **BACKWARD INCOMPATIBLE:** Require Resolwe 12.x
```
Comment (JureZmrzlikar) [inline docs/CHANGELOG.rst]: It would be cleaner to put these changes in a separate commit.

File: resolwe_bio/tests/processes/test_support_processors.py (hunk 7 above)
Comment (jkokosar) [reply resolwe_bio/tests/processes/test_support_processors.py]: I have improved this slightly. A function to assert a file within a `dir` field would be of great use here.

File: resolwe_bio/processes/alignment/star.yml [hunk 9]
```diff
@@ -387,8 +454,21 @@
...
       re-progress 0.8
-      samtools index Aligned.sortedByCoord.out.bam "${READS_NAME}.bam.bai"
-      re-checkrc
+
+      # if alignemnt job produces empty .bam file, the file needs to be re-sorted
+      # for it to be successfully indexed by samtools.
+      if [[ $(samtools view -c Aligned.sortedByCoord.out.bam) == 0 ]]
+      then
+        samtools sort Aligned.sortedByCoord.out.bam > Aligned.sortedByCoord_temp.out.bam
+        re-checkrc
```
Comment (mzganec) [inline resolwe_bio/processes/alignment/star.yml]: Add a message.
